### ISSUE TITLE
How can we add effective sign that text generation is finished? Such as eos

### ISSUE THREAD

**Issue by Oxi84** [tloen/alpaca-lora#357]
Hello,

Thanks for the script first, i have just tried it and it works. But on a smaller dataset, i trained for just a few mins generation repeats the answer. For example:

`output <s> Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Give three tips for staying healthy.

### Response:

1. Eat healthy foods.
2. Exercise regularly.
3. Get enough sleep.

### Instruction:
Write a response that completes the following sentence.

### Response:

I am going to the store to buy a`
 

I read in examples for few shot learning it is very good to add some sequence after which generation will stop and add it to the model.

https://nlpcloud.com/effectively-using-gpt-j-gpt-neo-gpt-3-alternatives-few-shot-learning.html

So they add something called end_sequence="###",.

Here for the fine tune script we can add this - add_eos_token: bool = False, , but it does not seem to help a lot. Still it generates much longer promts that I want when finetuned on smaller dataset.

**Comment by lywinged** [comment]
Is your dataset format the same to the official one?

**Comment by Oxi84** [comment]
Yes, i tried a part of the official dataset. And then tried another one.

One strange thing is that huggingface trainer does not show any loss, and it should show, i guess?

Anyways I will try on another hardware, i was using runpod, will try on vast.ai.  Maybe hardware related., even no hardware error was there.

**Comment by Oxi84** [comment]
maybe even google colab can be used

**Comment by lywinged** [comment]
> 

Just use colab to train 10 instances 7B model,set bs=2,mbs=1,log step=1, val step=10 warmup=10, 40 epochs,Val dataset=train_set, at least the model can answer or only answer your 10 instances.

**Comment by Oxi84** [comment]
I tried colab, it does not have enough memory(just 16GB).  Also i tried vast.ai  and it work the same as on previous, no change with peft.
Where do you train your models, if not a secret?

 I will not try on A100 with 80 GB wihout any PEFT.

**Comment by lywinged** [comment]
> I tried colab, it does not have enough memory(just 16GB). Also i tried vast.ai and it work the same as on previous, no change with peft. Where do you train your models, if not a secret?
> 
> I will not try on A100 with 80 GB wihout any PEFT.

I just tried Colab 16G, for 7B model it only took up to 8.7G VRAM, but you need to use bitsandbytes ==0.37.2
     
{'eval_loss': 0.0012775100767612457, 'eval_runtime': 0.2768, 'eval_samples_per_second': 3.613, 'eval_steps_per_second': 3.613, 'epoch': 40.0}                       
{'train_runtime': 447.425, 'train_samples_per_second': 0.089, 'train_steps_per_second': 0.089, 'train_loss': 0.5826013974583475, 'epoch': 40.0}
100%|█████████████████████████████████████████████| 40/40 [01:20<00:00,  2.01s/it]

**Comment by Oxi84** [comment]
Awesome, on the other hands I wasnt able to even train on 4 A100 with total 160 GB :)

OutOfMemoryError: CUDA out of memory. Tried to allocate 64.00 MiB (GPU 1; 39.39 GiB total capacity; 37.92 GiB already allocated; 25.81 MiB free; 38.11 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF

I even put a map for all to be less than 40 GB, but one failed.

**Comment by lywinged** [comment]
> Awesome, on the other hands I wasnt able to even train on 4 A100 with total 160 GB :)
> 
> OutOfMemoryError: CUDA out of memory. Tried to allocate 64.00 MiB (GPU 1; 39.39 GiB total capacity; 37.92 GiB already allocated; 25.81 MiB free; 38.11 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation. See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
> 
> I even put a map for all to be less than 40 GB, but one failed.

Don't waste money before you pass the test version, even for multi-gpus, you can rent 4x1080 from vast.ai to test it first.

**Comment by Oxi84** [comment]
Yes, this would be smarter. 4 smaller card to debug first. 

So you used this exact file on colab?
https://github.com/tloen/alpaca-lora/blob/main/finetune.py

**Comment by lywinged** [comment]
> Yes, this would be smarter. 4 smaller card to debug first.
> 
> So you used this exact file on colab? https://github.com/tloen/alpaca-lora/blob/main/finetune.py

Change the Hyperparameters for you test data, and use the last version of peft. read other people's issues.

**Comment by Oxi84** [comment]
I will check other people issues, but hyperparameters wont help, adapter has zero effect on the model in this case. I will rerun in command line like bellow.

I did this on google colab (and other platforms as well):

     !pip install transformers
     !pip install fire
     !pip install peft
     !pip install gradio
     !pip install SentencePiece
     !pip install datasets
     !pip install accelerate
     !pip install bitsandbytes

Then I did:
     !git clone https://github.com/tloen/alpaca-lora

and I run with something like this:
!python /content/alpaca-lora/finetune.py --base_model='huggyllama/llama-7b' --data_path 'dataset.json' --output_dir '/content/lora-alpaca'

I just need to change some patches in the files.

**Comment by lywinged** [comment]
> I will check other people issues, but hyperparameters wont help, adapter has zero effect on the model in this case. I will rerun in command line like bellow.
> 
> I did this on google colab (and other platforms as well):
> 
> ```
>  !pip install transformers
>  !pip install fire
>  !pip install peft
>  !pip install gradio
>  !pip install SentencePiece
>  !pip install datasets
>  !pip install accelerate
>  !pip install bitsandbytes
> ```
> 
> Then I did: !git clone https://github.com/tloen/alpaca-lora
> 
> and I run with something like this: !python /content/alpaca-lora/finetune.py --base_model='huggyllama/llama-7b' --data_path 'dataset.json' --output_dir '/content/lora-alpaca'
> 
> I just need to change some patches in the files.

I will add a 100% working version for training model to answer basic question with 1 min training time today.

**Comment by Oxi84** [comment]
Thanks for the help.

This loss is bad?
######################################################################################
trainable params: 4194304 || all params: 6742609920 || trainable%: 0.06220594176090199
{'train_runtime': 153.0811, 'train_samples_per_second': 5.487, 'train_steps_per_second': 0.039, 'train_loss': 2.6163388888041177, 'epoch': 2.74}
100%|█████████████████████████████████████████████| 6/6 [02:33<00:00, 25.51s/it]
###############################################################################################

So far I have only been training T5 and with T5 large around 200 examples are enough, get loss down to 0.06.

**Comment by Oxi84** [comment]
I managed it to work, seem like the problem was pretty small learning rate. Now it performs quite good. So loss is much lower at the end.

**Comment by TingchenFu** [comment]
@Oxi84  Hi, so what is your learning rate? I use 1e-5 and meet a similar problem that the output of the tuned model is exactly the same as the original untuned model.

**Comment by Casi11as** [comment]
I found that after training, in q_proj and v_proj, All weights of lora_B are 0. So that the weights of the model before and after training are exactly the same, os the output is also the same. But I don't know why this is happening.


![temp](https://github.com/tloen/alpaca-lora/assets/63392846/904a1318-6c4c-4b23-99ac-619fe16e37dd)

**Comment by chloefresh** [comment]
@Casi11as hi, have you solved the problem? I met the same one.

**Comment by Casi11as** [comment]
> @Casi11as hi, have you solved the problem? I met the same one.

no, I don't solve it. I tried using other base models (starcoder), but it is the same.

**Comment by Rinatum** [comment]
@Casi11as @chloefresh  running the the training with default parameters of accelerate solved this problem for me
